### PR TITLE
QWMA diff algo

### PR DESCRIPTION
**Pull request by fireice-uk**
- Cleaned up confusion regarding `DIFFICULTY_WINDOW_V3` vs `DIFFICULTY_BLOCKS_COUNT_V3` in the current algo.

- Added a new quadratic algo that tends to favour newer results.

### CODE REVIEW

**Comment by fireice-uk** [comment]
Integrated the algo tested on our harness, ready to merge

File: src/cryptonote_basic/difficulty.cpp [hunk 1]
```diff
@@ -320,4 +320,108 @@ difficulty_type next_difficulty_v3(const std::vector<std::uint64_t> &timestamps,
...
+	uint64_t maxValid = timestamps[N];
+	for(size_t i = 1; i < N; i++)
+	{
+		/* 
+		 * Mask timestamp if it is smaller or equal to last valid timestamp
+		 * or if it is larger or equal to largest timestamp
+		 */
+		if(timestamps[i] <= findLastValid(timestamps, i) || timestamps[i] >= maxValid)
+		{
+			if(i != 1)
```
Comment (psychocrypt) [inline src/cryptonote_basic/difficulty.cpp]: this must be fixed also in the harness

File: src/cryptonote_basic/difficulty.cpp [hunk 2]
```diff
@@ -320,4 +320,108 @@ difficulty_type next_difficulty_v3(const std::vector<std::uint64_t> &timestamps,
 
 	return next_D;
 }
+
+//Find non-zero timestamp with index smaller than i
+inline uint64_t findLastValid(const std::vector<uint64_t>& timestamps, size_t i)
+{
+	while(--i > 1)
```
[the remark this answers is not in the document]
Comment (fireice-uk) [reply src/cryptonote_basic/difficulty.cpp]: Addressed and added test.

File: src/cryptonote_basic/difficulty.cpp (hunk 1 above)
Comment (fireice-uk) [reply src/cryptonote_basic/difficulty.cpp]: Done

**Pull request merged**